core/lib/Cardano/Address/Script.hs

```diff
@@ -258,77 +321,201 @@ foldScript fn zero = \case
+        when (length (omitTimelocks script) < fromIntegral m) $ Left ListTooSmall
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (redundantTimelocks script) $ Left RedundantTimelocks
+        traverse_ recommendedValidation script
 
     ActiveFromSlot _ -> pure ()
 
     ActiveUntilSlot _ -> pure ()
   where
-    hasDuplicate xs = do
+    hasDuplicate xs =
         length sigs /= length (L.nub sigs)
       where
         sigs = [ sig | RequireSignatureOf sig <- xs ]
+    hasTimelocks = \case
+        ActiveFromSlot _ -> True
+        ActiveUntilSlot _ -> True
+        _ -> False
+    redundantTimelocks xs = case L.filter hasTimelocks xs of
+        [] -> False
+        [_] -> False
+        [ActiveFromSlot s1, ActiveUntilSlot s2] -> s2 <= s1
+        [ActiveUntilSlot s2, ActiveFromSlot s1] -> s2 <= s1
+        _ -> True
+    omitTimelocks = filter (not . hasTimelocks)
+--
+-- ScriptTemplate validation
+--
+
+-- | Validate a 'ScriptTemplate', semantically
+--
+-- @since 3.2.0
+validateScriptTemplate
+    :: ValidationLevel
+    -> TxValidity
+    -> ScriptTemplate
+    -> Either ErrValidateScriptTemplate ()
+validateScriptTemplate level interval (ScriptTemplate cosigners' script) = do
+    when (Map.size cosigners' == 0) $ Left NoCosigner
+    when (L.length (L.nub $ Map.elems cosigners') /= Map.size cosigners') $
+        Left DuplicateXPubs
+    let allCosigners = Set.fromList $ foldScript (:) [] script
+    let unknownCosigners =
+            allCosigners `difference` Set.fromList (Map.keys cosigners')
+    unless (Set.null unknownCosigners) $ Left UnknownCosigner
+    let unusedCosigners =
+            Set.fromList (Map.keys cosigners') `difference` allCosigners
+    unless (Set.null unusedCosigners) $ Left UnusedCosigner
+    mapLeft WrongScript $ do
+        unless (requiredValidation interval script)
+            $ Left LedgerIncompatible
+        when (level == RecommendedValidation ) $
+            recommendedValidation script
 
 -- | Possible validation errors when validating a script
 --
 -- @since 3.0.0
 data ErrValidateScript
-    = EmptyList
+    = LedgerIncompatible
+    | EmptyList
     | ListTooSmall
     | MZero
     | DuplicateSignatures
     | WrongKeyHash
     | Malformed
+    | RedundantTimelocks
     deriving (Eq, Show)
 
--- | Pretty-print a validation error.
+-- | Possible validation errors when validating a script
+--
+-- @since 3.2.0
+data ErrValidateScriptTemplate
+    = WrongScript ErrValidateScript
+    | DuplicateXPubs
+    | UnknownCosigner
+    | NoCosigner
+    | UnusedCosigner
+    deriving (Eq, Show)
+
+-- | Pretty-print a script validation error.
 --
 -- @since 3.0.0
 prettyErrValidateScript
     :: ErrValidateScript
     -> String
 prettyErrValidateScript = \case
+    LedgerIncompatible ->
+        "The script is ill-formed and is not going to be accepted by ledger."
     EmptyList ->
         "The list inside a script is empty."
```

Comment:
Note that with the `omitTimelock` above, this message may be a bit confusing and should be reworked. "The list inside a script is empty or only contains timelocks (which is not recommended)."

Reply:
refactored errors, also added (I hope) better messaging. see HEAD